cmd/egress-cni-plugin/cni/cni4_test.go

```diff
@@ -0,0 +1,201 @@
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package cni
+
+import (
+	"fmt"
+	"net"
+	"testing"
+
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/stretchr/testify/assert"
+
+	. "github.com/aws/amazon-vpc-cni-k8s/cmd/egress-cni-plugin/share"
+	mock_iptables "github.com/aws/amazon-vpc-cni-k8s/pkg/iptableswrapper/mocks"
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/utils/logger"
+
+	mock_netlinkwrapper "github.com/aws/amazon-vpc-cni-k8s/pkg/netlinkwrapper/mocks"
+
+	mock_nswrapper "github.com/aws/amazon-vpc-cni-k8s/pkg/nswrapper/mocks"
+
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/golang/mock/gomock"
+
+	mock_ipamwrapper "github.com/aws/amazon-vpc-cni-k8s/pkg/ipamwrapper/mocks"
+	mock_procsyswrapper "github.com/aws/amazon-vpc-cni-k8s/pkg/procsyswrapper/mocks"
+	mock_veth "github.com/aws/amazon-vpc-cni-k8s/pkg/vethwrapper/mocks"
+)
+
+func TestCmdAddEgressV4(t *testing.T) {
+	context := setupAddContextV4(gomock.NewController(t))
+
+	expectIptablesRules := []string{
+		"nat CNI-E4 -d 224.0.0.0/4 -j ACCEPT -m comment --comment unit-test-comment",
+		"nat CNI-E4 -j SNAT --to-source 192.168.0.23 -m comment --comment unit-test-comment --random-fully",
+		"nat POSTROUTING -s 169.254.172.100 -j CNI-E4 -m comment --comment unit-test-comment"}
+	var actualIptablesRules []string
+
+	expectRouteDel := []string{"route del: {Ifindex: 2 Dst: 169.254.172.0/22 Src: <nil> Gw: <nil> Flags: [] Table: 0}"}
+	var actualRouteDel []string
+
+	expectRouteAdd := []string{
+		"route add: {Ifindex: 2 Dst: 169.254.172.1/32 Src: 169.254.172.100 Gw: <nil> Flags: [] Table: 0}",
+		"route add: {Ifindex: 2 Dst: 169.254.172.0/22 Src: 169.254.172.100 Gw: 169.254.172.1 Flags: [] Table: 0}",
+		"route add: {Ifindex: 100 Dst: 169.254.172.100/32 Src: <nil> Gw: <nil> Flags: [] Table: 0}"}
+	var actualRouteAdd []string
+
+	// setup the mock EXPECT
+	err := SetupAddExpectV4(context, &actualIptablesRules, &actualRouteAdd, &actualRouteDel)
+	assert.Nil(t, err)
+
+	// run egress plugin ADD action
+	err = CmdAddEgressV4(&context)
+	assert.Nil(t, err)
+
+	// confirm all iptables chain/rule are created
+	assert.EqualValuesf(t, expectIptablesRules, actualIptablesRules, "expected iptables chain/rule are not created")
+
+	// confirm automatically added route is removed
+	assert.EqualValuesf(t, expectRouteDel, actualRouteDel, "route are expected to removed")
+
+	// confirm routes are added in container and node
+	assert.EqualValuesf(t, expectRouteAdd, actualRouteAdd, "route are expected to added")
+
+	// the unit test write some output string not ends with '\n' and this cause go runner unable to interpret that a test was run.
+	// Adding a newline, keeps a clean output
+	fmt.Println()
+
```

Comment:
This println must be here otherwise unit testing will be treated `not run` even it successfully run.

The reason is that our plugin write some strings to stdout and it does not ends with a newline, this cause test runner could not find the string `---- passed`.

Reply:
I meant the empty newline, but I am wondering if we can find that write and add a newline to it, as all of our logging should use newlines

Reply:
oh, you mean empty new line in code, I just removed it.

The code that generate the output without newline - it is containernetworking/cni package

```
return types.PrintResult(c.Result, c.NetConf.CNIVersion)
```